### PR TITLE
Add Auto-Doc Service

### PR DESCRIPTION
**Pull request by Coolroo**
Added functionality:
- Added new documentation insertion functionality, which deletes old docstrings
- Added change tracker service which finds changes in the file being updated
- Added function highlighter which highlights updated functions
- Added auto-doc functionality which will automatically document changed functions
- Added additional configurations which change the default behaviour of the auto-doc service
- Added tags (@trelent-auto, @trelent-highlight, @trelent-ignore) which can modify the default behaviour of the autodoc-service for a specified function

### CODE REVIEW

**Comment by calum-bird** [comment]
Looks like the test fails, any ideas?

File: package.json
Comment (calum-bird) [inline package.json:1]: Going to update the descriptions here to be a bit more clear, and have both `mode` and `changeThreshold` under `trelent.autodoc`.

File: src/extension.ts [hunk 1]
```diff
@@ -19,7 +22,7 @@ var publicMPToken = "6a946c760957a81165973cc1ad5812ec";
 // your extension is activated the very first time the command is executed
 export async function activate(context: vscode.ExtensionContext) {
   // Setup our Telemetry Service
-  var telemetryService = new TelemetryService(publicMPToken);
+  var telemetryService = createTelemetryService(publicMPToken);
```
Comment (calum-bird) [inline src/extension.ts:25]: What happened here? This seems like an unnecessary change, and moves us away from the provider architecture.

File: package.json [hunk 2]
```diff
@@ -137,9 +137,51 @@
...
+            "Enable Per-Function",
+            "Enable Globally",
+            "Maintain Docstrings"
+          ],
+          "enumDescriptions": [
+            "If set to 'Enable Per-Function', the auto-doc service will only notify of changes for functions with the tag @trelent-highlight.",
+            "If set to 'Enable Globally', the auto-doc service will be enabled for all functions, functions titled with the tag @trelent-ignore will not be highlighted.",
+            "If set to 'Maintain Docstrings', the auto-doc service will automatically document all functions automatically"
+          ],
+          "description": "Defines the behaviour of the auto-doc service."
```
Comment (calum-bird) [inline package.json]: Going to update the descriptions here to be a bit more clear, and have both mode and changeThreshold under trelent.autodoc.

File: src/extension.ts [hunk 3]
```diff
@@ -31,8 +34,10 @@ export async function activate(context: vscode.ExtensionContext) {
   var authService = new AuthenticationService(context, telemetryService);
 
   // Setup our CodeParser service
-  const pending = new CodeParserService(context);
-  var codeParserService = await pending;
+  var codeParserService = await createCodeParserService(
+    context,
+    telemetryService
+  );
```
Comment (calum-bird) [inline src/extension.ts:40]: Again, not sure if I'm a fan of having the service contain a function to create itself? Why not use a constructor as with other services?

File: src/services/telemetry.ts [hunk 4]
```diff
@@ -84,3 +84,16 @@ function getTelemetrySettings(): number {
...
+export let createTelemetryService = (mixpanelToken: string) => {
+  if (!service) {
+    service = new TelemetryService(mixpanelToken);
+  }
+  return service;
+}
+
+export let getTelemetryService = () => {
+  return service;
+}
```
Comment (calum-bird) [inline src/services/telemetry.ts:99]: Again, confused by the necessity of this. Only one telem service should ever exist in extension.ts if we use a constructor, why do we need this check?

File: tsconfig.extension.json [hunk 5]
```diff
@@ -3,7 +3,7 @@
     "module": "commonjs",
     "target": "es6",
     "outDir": "build",
-    "lib": ["es6", "dom"],
+    "lib": ["es6", "dom", "es2019"],
```
Comment (calum-bird) [inline tsconfig.extension.json:6]: Is this okay for all vscode versions? What necessitated this?

**Review by calum-bird — CHANGES_REQUESTED** [review]
Lets fucking go! So hyped for this to go live. I left some comments on areas I was confused by, but other than those it looks great. I'll be pushing a few commits with some changes here and there, mostly aesthetic.

LMK about those comments I left, guessing I'm just missing something.

File: tsconfig.extension.json (hunk 5 above)
Comment (Coolroo) [reply tsconfig.extension.json:6]: I had an issue and had to add this, don't remember what it was, maybe it resolved itself? I will delete it and see if it still works

File: src/services/telemetry.ts (hunk 4 above)
Comment (Coolroo) [reply src/services/telemetry.ts:99]: If other services need it, we'd need to pass it in the constructor. This way, instead of passing it to all the services that need it, we can just call getTelemetryService instead, which imo is better functionality wise. EG: say a service that is a child of another service, which is a child of another service, which is defined in the extension needs it, we'd need to add it to the constructor of all the classes in between. This way we can just call a function and be done with it

File: src/extension.ts (hunk 3 above)
Comment (Coolroo) [reply src/extension.ts:40]: This makes it so we can treat the service as a singleton. The issue we were having with tests was that multiple code parser services were trying to be created, and they were all trying to register commands. This way there will only ever be one instance of the service, and we can fetch it using the getter method

File: package.json (hunk 2 above)
Comment (Coolroo) [reply package.json]: Agreed!

File: src/extension.ts (hunk 1 above)
Comment (Coolroo) [reply src/extension.ts:25]: This incorporates a singleton architecture which allows us to access services more fluidly between services

File: src/extension.ts (hunk 3 above)
Comment (calum-bird) [reply src/extension.ts:40]: I think service might just be the wrong name here - we aren't really making a service, more an instance of a class where a singleton makes sense. Either way, I'm much more comfortable doing this here with the code parser than I am with the telemetry service.

File: src/extension.ts (hunk 1 above)
Comment (calum-bird) [reply src/extension.ts:25]: Right, let's discuss tomorrow and walk through this a bit more.

File: tsconfig.extension.json (hunk 5 above)
Comment (calum-bird) [reply tsconfig.extension.json:6]: "It works on my machine" lmao

File: src/services/telemetry.ts (hunk 4 above)
Comment (calum-bird) [reply src/services/telemetry.ts:99]: Gotcha. Again, naming issue at this point. We'll keep it as a singleton but rename along with everything else that is refactored.

File: src/autodoc/DocstringInsertService.ts [hunk 6]
```diff
@@ -166,23 +166,12 @@ export default class DocstringInsertService {
...
-          return;
-        }
-
-        // Check if any
-        offsetVal += await this.documentFunctions(
-          autoFunctions,
-          editor,
-          document
-        );
-      }, 500);
```
Comment (calum-bird) [inline src/autodoc/DocstringInsertService.ts]: Darn, I really liked this approach! It would have made a fun challenge for the next intern - "go find where we add a slight delay to the generation"

Comment (Coolroo) [reply src/autodoc/DocstringInsertService.ts]: lmao

**Review by calum-bird — APPROVED** [review]
Lets gooo
